clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3404,21 +3404,23 @@ void Sema::MarkDevice(void) {
 // SYCL device specific diagnostics implementation
 // -----------------------------------------------------------------------------
 
-Sema::SemaDiagnosticBuilder Sema::SYCLDiagIfDeviceCode(SourceLocation Loc,
-                                                       unsigned DiagID) {
+Sema::SemaDiagnosticBuilder
+Sema::SYCLDiagIfDeviceCode(SourceLocation Loc, unsigned DiagID,
+                           DeviceDiagnosticReason Reason) {
   assert(getLangOpts().SYCLIsDevice &&
          "Should only be called during SYCL compilation");
   FunctionDecl *FD = dyn_cast<FunctionDecl>(getCurLexicalContext());
-  SemaDiagnosticBuilder::Kind DiagKind = [this, FD] {
+  SemaDiagnosticBuilder::Kind DiagKind = [this, FD, Reason] {
     if (DiagnosingSYCLKernel)
       return SemaDiagnosticBuilder::K_ImmediateWithCallStack;
     if (!FD)
       return SemaDiagnosticBuilder::K_Nop;
-    if (getEmissionStatus(FD) == Sema::FunctionEmissionStatus::Emitted)
+    if (getEmissionStatus(FD) == Sema::FunctionEmissionStatus::Emitted &&
+        (getEmissionReason(FD) & Reason) != Sema::DeviceDiagnosticReason::None)
       return SemaDiagnosticBuilder::K_ImmediateWithCallStack;
     return SemaDiagnosticBuilder::K_Deferred;
```

Comment:
My point is this case... getEmissionReason seems to need a 'there is no known reason to emit this' flag vs 'all' vs 'none'.  

1 'All' needs to mean "There are multiple reasons to emit this".
2 'None' needs to mean "There is NO reason to emit this (or to ever emit this!)'.
3 and an 'Unknown' state that means, "I don't know yet".

The latter means 'Defer' (and in the deferred emission, is an assert-worthy-issue).  Currently we are returning 'All' to mean both 1 and 3.  I suspect it needs another state outside of the bitmask for 'not sure'.